orttraining/orttraining/core/optimizer/shape_optimizer.cc

```diff
@@ -0,0 +1,355 @@
+      } else {
+        // Fill with -1 for symbolic dimension.
+        dim_values.push_back(-1);
+      }
+    }
+
+    if (!has_concrete_dim) {
+      SO_LOG_DEBUG_INFO(logger, "No concrete dimension found, don't need try further." + node->Name());
+      continue;
+    }
+
+    InlinedVector<Node*> nodes_to_remove;
+    TensorShapeVector fold_values;
+    // Short path - check if the shape node can be constant folded.
+    if (CanShapeNodeBeReplacedWithConstant(*node, dim_values, fold_values)) {
+      SO_LOG_DEBUG_INFO(logger, "Shape node can be constant folded." + node->Name());
+      UpdateNodeArgToConstant(graph, node->MutableOutputDefs()[0], fold_values);
+      nodes_to_remove.push_back(node);
+    } else {
+      // Check consumers of Shape node, try best effort to constant fold them if possible.
+      // Currently support Gather and Slice in some cases.
+      auto p_ip_node = node->OutputNodesBegin();
+      const auto p_ip_node_end = node->OutputNodesEnd();
+      InlinedHashSet<const Node*> visited_nodes;
+      while (p_ip_node != p_ip_node_end) {
+        if (visited_nodes.find(&(*p_ip_node)) != visited_nodes.end()) {
+          // Already handled, skip the node.
+          ++p_ip_node;
+          continue;
+        }
+
+        auto& output_node = const_cast<Node&>(*p_ip_node);
+        visited_nodes.insert(&output_node);
+        ++p_ip_node;
+
+        NodeArg* data_input = output_node.MutableInputDefs()[0];
+        // Skip when shape is not used as sliced data.
+        if (data_input != node->MutableOutputDefs()[0]) {
+          continue;
+        }
+
+        TensorShapeVector slice_fold_values;
+        if (graph_utils::IsSupportedOptypeVersionAndDomain(output_node, "Slice", {10, 11, 13}) &&
+            CanSliceNodeBeReplacedWithConstant(graph, output_node, dim_values, slice_fold_values)) {
+          SO_LOG_DEBUG_INFO(logger, "Slice node can be constant folded." + output_node.Name());
+          UpdateNodeArgToConstant(graph, output_node.MutableOutputDefs()[0], slice_fold_values);
+          nodes_to_remove.push_back(&output_node);
+          continue;
+        }
+
+        int gather_output_rank = 0;
+        TensorShapeVector gather_fold_values;
+        if (graph_utils::IsSupportedOptypeVersionAndDomain(output_node, "Gather", {1, 11, 13}) &&
+            CanGatherNodeBeReplacedWithConstant(graph, output_node, dim_values, gather_fold_values,
+                                                gather_output_rank)) {
+          SO_LOG_DEBUG_INFO(logger, "Gather node can be constant folded." + output_node.Name());
+          UpdateNodeArgToConstant(graph, output_node.MutableOutputDefs()[0], gather_fold_values, gather_output_rank == 0);
+          nodes_to_remove.push_back(&output_node);
+          continue;
+        }
+      }
+    }
+
+    for (Node* node_to_remove : nodes_to_remove) {
+      // Remove single-output node chain for inputs of the node
+      auto p_ip_node = node_to_remove->InputNodesBegin();
```

Comment:
nit: the code logic should be common in other graph transformers? may be a put it as an graph util function?